// @b360/phpmailer/index.js
'use strict';

var config = require('@b360/config');

// to add clients - register
exports.sendPHPMail = async (data) => {
	console.log(data)
	let result = [];
	let msg = [];
	// validations 
	if(data['mail_to'] == '')
		msg['mail_to'] = 'mail_to Required';

	if(data['mail_to_name']=='')
		msg['mail_to_name'] = 'mail_to_name Required';

	if(data['mail_subject']=='')
		msg['mail_subject'] = 'mail_subject Required';

	if(data['mail_message']=='')
		msg['mail_message'] = 'mail_message Required';

	if(data['setting']=='')
		msg['setting'] = 'setting Required';

	if(data['smtp_host']=='')
		msg['smtp_host'] = 'smtp_host Required';

	if(data['smtp_port']=='')
		msg['smtp_port'] = 'smtp_port Required';

	// if(data['g-recaptcha-token']!='')
	// 	msg['g-recaptcha-token']=='g-recaptcha-token';

	if(data['debug']=='')
		msg['debug'] = 'debug Required';

	if(data['email']=='')
		msg['email'] = 'email Required';

	if(data['password']=='')
		msg['password'] = 'password Required';

	if(data['mail_from_email']=='')
		msg['mail_from_email'] = 'mail_from_email Required';

	if(data['mail_from_name']=='')
		msg['mail_from_name'] = 'mail_from_name Required';


	let msgLength = Object.keys(msg).length;
	if(!msgLength){

		var myHeaders = new Headers();

		var formdata = new FormData();
		formdata.append("mail_to", data['mail_to']);
		formdata.append("mail_to_name", data['mail_to_name']);
		formdata.append("mail_subject", data['mail_subject']);
		formdata.append("mail_message", data['mail_message']);
		formdata.append("setting", data['setting']);
		formdata.append("smtp_host", data['smtp_host']);
		formdata.append("smtp_port", data['smtp_port']);
		formdata.append("debug", data['debug']);
		formdata.append("email", data['email']);
		formdata.append("password", data['password']);
		formdata.append("mail_from_email", data['mail_from_email']);
		formdata.append("mail_from_name", data['mail_from_name']);
		if(data['g_recaptcha_token'] != ''){
			formdata.append("g-recaptcha-token", data['g_recaptcha_token']);
		}

		console.log(formdata);
		console.log(Object.entries(formdata))


		var requestOptions = {
			method: 'POST',
			headers: myHeaders,
			body: formdata,
			redirect: 'follow'
		};

		return fetch('https://phpmailer.gisolutions.nz/index.php', requestOptions)
		.then(response => response.text())
		.then(result => {
			result = JSON.parse(result);
			console.log(result.status);
			console.log(result.message);
			return result;
		})
		.catch((error) => {
			console.log('error', error);
			result['status'] = false;
			result['message'] = error;
			return result;
		});

	}else{
		result['status'] = false;
		result['message'] = msg;
		result['errors'] = msg;
		return result;
	}	
}
